angelsdev-unit-test/unit-tests/unit-test-009.lua

```diff
@@ -35,6 +36,68 @@ local function has_recipe(recipe_filters, recipes_to_ignore)
   end
 end
 
+local function try_find_character_for(recipe)
+  local entity_filters = {}
+  table.insert(entity_filters, {filter = "name", name = "character", mode = "and"})
+  table.insert(entity_filters, {filter = "hidden", invert = true, mode = "and"})
+  table.insert(entity_filters, {filter = "type", type = "character", mode = "and"})
+  
+  local entities = game.get_filtered_entity_prototypes(entity_filters)
+
+  for entity_name, entity in pairs(entities) do
+    for category_name, _ in pairs(entity.crafting_categories) do
+      if recipe.category == category_name then
+        return true
+      end
+    end
+  end
+end
+
+local function try_find_entity_for(recipe)
+  -- Count ingredients and products in recipe
+  local item_ingredient_count = 0
+  local fluid_ingredient_count = 0
+  local fluid_product_count = 0
+
+  for _, ingredient in pairs(recipe.ingredients) do
+    if ingredient.type == "item" then
+      item_ingredient_count = item_ingredient_count + 1
+    elseif ingredient.type == "fluid" then
+      fluid_ingredient_count = fluid_ingredient_count + 1
+    end
+  end
+
+  for _, product in pairs(recipe.products) do
+    if product.type == "fluid" then
+      fluid_product_count = fluid_product_count + 1
+    end
+  end
+
+  -- Try find an entity that can craft this recipe
+  local entity_filters = {}
+  table.insert(entity_filters, {filter = "type", type = "assembling-machine", mode = "or"})
+  table.insert(entity_filters, {filter = "crafting-category", crafting_category = recipe.category, mode = "and"})
+  table.insert(entity_filters, {filter = "hidden", invert = true, mode = "and"})
+
+  table.insert(entity_filters, {filter = "type", type = "furnace", mode = "or"})
+  table.insert(entity_filters, {filter = "crafting-category", crafting_category = recipe.category, mode = "and"})
+  table.insert(entity_filters, {filter = "hidden", invert = true, mode = "and"})
+
+  local entity_prototypes = game.get_filtered_entity_prototypes(entity_filters)
```

Comment:
I believe you forgot to add another set of filters for type = `rocket-silo`, they also accept recipes...

Reply:
```suggestion
  -- Try find an entity that can craft this recipe
  local entity_filters = {}
  table.insert(entity_filters, {filter = "type", type = "assembling-machine", mode = "or"})
  table.insert(entity_filters, {filter = "crafting-category", crafting_category = recipe.category, mode = "and"})
  table.insert(entity_filters, {filter = "hidden", invert = true, mode = "and"})
  
  table.insert(entity_filters, {filter = "type", type = "furnace", mode = "or"})
  table.insert(entity_filters, {filter = "crafting-category", crafting_category = recipe.category, mode = "and"})
  table.insert(entity_filters, {filter = "hidden", invert = true, mode = "and"})

  table.insert(entity_filters, {filter = "type", type = "rocket-silo", mode = "or"})
  table.insert(entity_filters, {filter = "crafting-category", crafting_category = recipe.category, mode = "and"})
  table.insert(entity_filters, {filter = "hidden", invert = true, mode = "and"})
  
  local entity_prototypes = game.get_filtered_entity_prototypes(entity_filters)
```